Sprint1/step 2

i0ji opened the pull request:
Modal refactored according to checklist and task.

i0ji committed:
0.0.1/b:step-1
yandex ui added

Commit:
header settings

Commit:
ingredients/builder setting

Commit:
ingredients/builder setting

Commit:
0.0.2/sprint1/step-1;
scrollbar added and set;
ingredient tabs;
builder markup;
ingredients JSON

Commit:
21.01.24/0.0.2/b:sprint1/step-1;
block size and general display set;
markup set;

Commit:
21.01.24/0.0.2/b:sprint1/step-1;
block size and general display set;
markup set;
ingredient group and card refactoring

i0ji committed:
22.01.24/0.0.3/b:sprint1/step-1
main markup done

i0ji committed:
22.01.24/0.0.3/b:sprint1/step-1
main markup done

i0ji committed:
minor edits

i0ji committed:
23.01.24/0.0.2/b:sprint1/step-1
minor edits;scrollbar sizing attempts;

Commit:
25.01.24/0.0.2/b:sprint1/step-1
last minor css edits;

Commit:
25.01.24/0.0.2/b:sprint1/step-1
last minor css edits; id to map method added in IngredientGroup

Commit:
25.01.24/0.1.0/b:sprint1/step-2
renamed blocks more logical;
fetch attempts;

Commit:
25.01.24/0.1.0/b:sprint1/step-2
got the date using a length check of the resulting array

i0ji committed:
26.01.24/0.1.1/b:sprint1/step-2
scss vars set; Modal start; multiply naming edits;

i0ji committed:
26.01.24/0.1.1/b:sprint1/step-2
constructing modal

Commit:
25.01.24/0.1.0/b:sprint1/step-2
Ingredient modal set;Started OrderDetails modal; multiple style touches; some vars renamed; small scroll refactoring

Commit:
0.1.3
orderModal done; small css edits

Commit:
0.1.3
scroll attempts

Commit:
27.01.24/0.1.3/
portal for modal attempts

Commit:
0.1.3
one modal done with closing on button, keyboard and outside click;

Commit:
0.1.3
one modal done with closing on button, keyboard and outside click;

Commit:
28.01.24/0.1.4/sprint1/step-2
two modals set and done;
scroll set in progress

Commit:
0.1.4/
API request, modals, scroll to ingredient groups, typing using TS and interfaces

Commit:
0.1.4/
small edits; assets -> images;

Commit:
0.1.5/
OrderDetails modal window refactoring according to the task

Commit:
0.1.5/
OrderDetails modal window refactoring according to the task

Commit:
0.1.5/
last small edits

Commit:
0.1.51/
props, types and interfaces optimization;
small code edits

andryxxa93 commented on src/App.tsx:
```diff
@@ -1,11 +1,57 @@
+    {/* ----- FETCHING DATA ON MOUNT ----- */
+    }
+    useEffect(() => {
+        fetch(ingredientsDataUrl)
```
**Можно лучше**

в дальнейшем добавятся ещё запросы к бэкенду и лучше вынести описание этих запросов в функции

Например:

```
function getIngredients() {
   return fetch(`${NORMA_API}/ingredients`)
    .then(проверка ответа)
}

```

и разместить их в отдельном файле  `utils/burger-api.js`, а не описывать запросы в самом компоненте

andryxxa93 commented on src/App.tsx:
```diff
@@ -1,11 +1,57 @@
+            .then(data => {
+                setIngredientsData(data.data)
+            })
+            .catch((err) => console.log(err.message))
```
**Можно лучше**
хорошо, что обрабатывается ошибка запроса к фзш api, но можно сделать еще лучше - сохранять в state и показывать пользователю ошибку в интерфейсе, чтобы он понимал что происходит

andryxxa93 commented on src/components/BurgerIngredients/BurgerIngredients.tsx:
```diff
@@ -0,0 +1,85 @@
+                    <IngredientGroup
+
+                        type="Булки"
+                        ingredients={ingredientsData.filter((elem) => elem.type === "bun")}
```
**Можно лучше**

Лучше в компоненте делать filter по массиву, а потом отображать полученный массив.

Например:

```
const buns = useMemo(data.filter((item) => item.type === 'bun'), [data]);
const mains = useMemo(data.filter((item) => item.type === 'main'), [data]);
const sauces = useMemo(data.filter((item) => item.type === 'sauce'), [data]);

return (
   ....................
   {buns.map(....................)}
   ..................
)
```

А не располагать фильтрацию в самом jsx

andryxxa93 reviewed (commented):
**Необходимо исправить**
- `useEffect` из `IngredientDetails ` должен быть внутри Modal и больше нигде.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

Commit:
0.1.5.2
API request separated into a separated file in src/utils; added error catching state;
refactoring of ingredient types filtering in BurgerIngs component using useMemo hook;

i0ji commented:
left to finalize
